plugins/ltac/tacexpr.mli

```diff
@@ -180,61 +180,58 @@ constraint 'a = <
 and 'a gen_tactic_expr =
   | TacAtom of ('a gen_atomic_tactic_expr) CAst.t
   | TacThen of
-      'a gen_tactic_expr *
-      'a gen_tactic_expr
+      ('a gen_tactic_expr *
+      'a gen_tactic_expr) CAst.t
```

Comment:
Most of the `CAst.t` you have added here should instead be done with the usual mutual pattern:
```ocaml
and 'a gen_tactic_expr_r = ... (* same as before *)
and 'a gen_tactic_expr = 'a gen_tactic_expr_f CAst.t
```

Reply:
I've tried several alternatives, but I get keep getting this error that I have trouble interpreting.

```
File "plugins/ltac/tacexpr.mli", line 181, characters 2-29:
Error: Constraints are not satisfied in this type.
       Type
       'a gen_tactic_expr_r
       should be an instance of
       < constant : 'b; dterm : 'c; level : 'd; name : 'e; pattern : 'f;
         reference : 'g; tacexpr : 'h; term : 'i >
       gen_tactic_expr_r
```

Reply:
You need to add the constraint too, as per OCaml.

 @ppedrot what is the point of using the `constrain` syntax here, shouldn't we maybe make this just a regular datatype?

Reply:
I didn't write this code, it was written by @aspiwack IIRC. I think that the motivation was to have a more robust code since it doesn't require to explicitly quantify over all the parameters in the type. Every time you change it (e.g. adding or removing a parameter) you have to do it everywhere. This style factorizes it in one single parameter.

Reply:
Well, you have to carry around the constraint anyways, so I'm not convinced this is not overkill IMHO; certainly it is not consistent with the rest of the codebase. I am tempted to try to use regular poly types to see if LOCs are lower.

[Already SerAPI does translate tacexpr to a type with 7 parameters and the constrain stuff confuses the hell out of the meta programming stuff]

Reply:
Yes, I'm responsible with changing type-level tuples into type-level records. The motivation was the same as for changing tuples into records in regular code: it's much easier to address into them.

I think it also made some of the code using it quite a bit less cluttered, and more robust to change (because you have to carry the constraint when you define types, but you don't need to speak of the entire tuple in the type of functions or something like this). Though that probably wasn't the main motivation. I don't remember for sure, it was almost ten years ago. (then again, I remember that at least one occurrence of me introducing a type-level record like this was a consequence of adding a type parameter to an already large tuple, maybe adding this type parameter in every signature was what convince me, I may be remembering some of it wrong :slightly_smiling_face: ).

Obviously, I'm not really directly concerned by this any more, so I will refrain to voice an opinion on which version is best.